indra/llcommon/lua_function.cpp

```diff
@@ -0,0 +1,894 @@
+// stack a Lua object corresponding to the passed LLSD object.
+void lua_pushllsd(lua_State* L, const LLSD& data)
+{
+    // might need 2 slots for array or map
+    luaL_checkstack(L, 2, nullptr);
+    switch (data.type())
+    {
+    case LLSD::TypeUndefined:
+        lua_pushnil(L);
+        break;
+
+    case LLSD::TypeBoolean:
+        lua_pushboolean(L, data.asBoolean());
+        break;
+
+    case LLSD::TypeInteger:
+        lua_pushinteger(L, data.asInteger());
+        break;
+
+    case LLSD::TypeReal:
+        lua_pushnumber(L, data.asReal());
+        break;
+
+    case LLSD::TypeBinary:
+    {
+        auto binary{ data.asBinary() };
+        std::memcpy(lua_newuserdata(L, binary.size()),
+                    binary.data(), binary.size());
+        break;
+    }
+
+    case LLSD::TypeMap:
+    {
+        // push a new table with space for our non-array keys
+        lua_createtable(L, 0, data.size());
+        for (const auto& pair: llsd::inMap(data))
+        {
+            // push value -- so now table is at -2, value at -1
+            lua_pushllsd(L, pair.second);
+            // pop value, assign to table[key]
+            lua_setfield(L, -2, pair.first.c_str());
+        }
+        break;
+    }
+
+    case LLSD::TypeArray:
+    {
+        // push a new table with space for array entries
+        lua_createtable(L, data.size(), 0);
+        lua_Integer key{ 0 };
+        for (const auto& item: llsd::inArray(data))
+        {
+            // push new array value: table at -2, value at -1
+            lua_pushllsd(L, item);
+            // pop value, assign table[key] = value
+            lua_rawseti(L, -2, ++key);
+        }
+        break;
+    }
+
+    case LLSD::TypeString:
+    case LLSD::TypeUUID:
+    case LLSD::TypeDate:
+    case LLSD::TypeURI:
+    default:
+    {
+        lua_pushstdstring(L, data.asString());
+        break;
+    }
+    }
+}
+
+/*****************************************************************************
+*   LuaState class
+*****************************************************************************/
+LuaState::LuaState(script_finished_fn cb):
+    mCallback(cb),
+    mState(nullptr)
+{
+    initLuaState();
+}
+
+void LuaState::initLuaState() 
+{
+    if (mState)
+    {
+        lua_close(mState);
+    }
+    mState = luaL_newstate();
+    luaL_openlibs(mState);
+    LuaFunction::init(mState);
+    // Try to make print() write to our log.
+    lua_register(mState, "print", LuaFunction::get("print_info"));
+    // We don't want to have to prefix require().
+    lua_register(mState, "require", LuaFunction::get("require"));
+}
+
+LuaState::~LuaState()
+{
+    // Did somebody call obtainListener() on this LuaState?
+    // That is, is there a LuaListener key in its registry?
+    auto listener{ getListener() };
+    if (listener)
+    {
+        // if we got a LuaListener instance, destroy it
+        auto lptr{ listener.get() };
+        listener.reset();
+        delete lptr;
```

Comment:
I guess it is magic: `LLInstanceTracker` magic. Arguably there should be an `LLInstanceTracker::destroy()` method or some such to encapsulate.

`LLInstanceTracker` stores a map of neutered `std::shared_ptr`s with [no-op deleters](https://github.com/secondlife/viewer/blob/release/luau-scripting/indra/llcommon/llinstancetracker.h#L248-L252). The point is that `LLInstanceTracker` is only to find subclass objects, not to manage their lifespan. That's what necessitates the logic in question.

Reply:
Got it. How do I have to bribe you, to add a bit of documentation about this behavior into the code?

Reply:
The best approach would be to add the aforementioned `LLInstanceTracker::destroy()` method, so that _its_ internal comments explain its implementation. I'll do that.

Reply:
https://github.com/secondlife/viewer/pull/1125